libvirtcollector/libvirtcollector.go

```diff
@@ -0,0 +1,238 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package libvirtcollector
+
+import (
+	wrapper "github.com/intelsdi-x/snap-plugin-collector-libvirt/libvirt"
+	"github.com/intelsdi-x/snap-plugin-utilities/config"
+	"github.com/intelsdi-x/snap/control/plugin"
+	"github.com/intelsdi-x/snap/control/plugin/cpolicy"
+	"github.com/intelsdi-x/snap/core"
+	"github.com/sandlbn/libvirt-go"
+)
+
+const (
+	// Name of plugin
+	Name = "libvirt"
+	// Type of plugin
+	Type = plugin.CollectorPluginType
+	// Vendor  prefix
+	Vendor = "intel"
+	// Plugin plugin name
+	Plugin = "libvirt"
+	// Version of plugin
+	Version = 10
+	// each metric starts with prefix "/intel/libvirt/<domain_id>"
+	lengthOfNsPrefix = 3
+)
+
+var nsTypes = struct {
+	cpu     []string
+	memory  []string
+	disk    []string
+	network []string
+	all     []string
+}{
+	cpu:     []string{"cputime"},
+	memory:  []string{"mem", "max", "swap_in", "swap_out", "major_fault", "min_fault", "unused", "available", "actual_balloon", "rss", "nr"},
+	disk:    []string{"wrreq", "rdreq", "wrbytes", "rdbytes"},
+	network: []string{"rxbytes", "rxpackets", "rxerrs", "rxdrop", "txbytes", "txpackets", "txerrs", "txdrop"},
+	all:     []string{"cpu", "memory", "disk", "network"},
+}
+
+// Meta declaration for plugin
+func Meta() *plugin.PluginMeta {
+	return plugin.NewPluginMeta(Name, Version, Type, []string{plugin.SnapGOBContentType}, []string{plugin.SnapGOBContentType})
+}
+
+// LibvirtCollector type
+type LibvirtCollector struct {
+}
+
+// Default qemu libvirt URI
+var defaultURI = "qemu:///system"
+
+// NewLibvirtCollector returns new instance of collector
+func NewLibvirtCollector() *LibvirtCollector {
+	return &LibvirtCollector{}
+
+}
+
+// CollectMetrics returns collected metrics
+func (p *LibvirtCollector) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricType, error) {
+	metrics := []plugin.MetricType{}
+	meta := make(map[string]string)
+
+	uri, err := config.GetConfigItem(mts[0], "uri")
+	if err != nil {
+		return nil, err
+	}
+	nova, err := config.GetConfigItem(mts[0], "nova")
+	if err != nil {
+		return nil, err
+	}
+	conn, err := libvirt.NewVirConnection(uri.(string))
+	if err != nil {
+		return nil, err
+	}
+	defer conn.CloseConnection()
+
+	ids, err := wrapper.GetInstances(conn)
+	if err != nil {
+		return metrics, err
+	}
+	for _, id := range ids {
+		if nova.(bool) {
+			meta, err = wrapper.GetNovaMetadata(id)
+			if err != nil {
+				return metrics, err
+			}
+		}
+
+		for _, mt := range filterNamespace("network", mts) {
+			ns := copyNamespace(mt)
+			ns[2].Value, err = id.GetName()
```

Comment:
It is just suggestion but consider creation of const  variables with descriptive names for parts of namespace to replace this `2` with  const variable. I think about something like `ns[nsNamePosition].Value` . Below there are similar cases for `ns[4]`, `n[5]`.